test/benchmark/schbench/smp1/run.sh

```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+# SPDX-License-Identifier: MPL-2.0
+
+/benchmark/bin/schbench -F 256 -n 5 -r 90
```

Comment:
Could you do me a favor by showing the text output of this command? I would like see the text output myself. This would be helpful to make a well-informed judgement on whether the proposed pattern matching is good.

Reply:
It appends the log every 10 seconds with the latest accumulated result and prints a final summary after 90 seconds.
<details>
  <summary>raw output</summary>

```sh
/benchmark/bin # ./schbench -F 256 -n 5 -r 90
setting worker threads to 8
Wakeup Latencies percentiles (usec) runtime 10 (s) (602 total samples)
          50.0th: 68         (177 samples)
          90.0th: 1570       (236 samples)
        * 99.0th: 74880      (54 samples)
          99.9th: 375296     (6 samples)
          min=1, max=375290
Request Latencies percentiles (usec) runtime 10 (s) (596 total samples)
          50.0th: 83072      (179 samples)
          90.0th: 190720     (237 samples)
        * 99.0th: 366080     (54 samples)
          99.9th: 415232     (5 samples)
          min=16122, max=415677
RPS percentiles (requests) runtime 10 (s) (11 total samples)
          20.0th: 57         (5 samples)
        * 50.0th: 58         (1 samples)
          90.0th: 61         (5 samples)
          min=55, max=61
current rps: 55.93
Wakeup Latencies percentiles (usec) runtime 20 (s) (1209 total samples)
          50.0th: 69         (354 samples)
          90.0th: 1834       (484 samples)
        * 99.0th: 85376      (108 samples)
          99.9th: 262656     (11 samples)
          min=1, max=375290
Request Latencies percentiles (usec) runtime 20 (s) (1202 total samples)
          50.0th: 81792      (360 samples)
          90.0th: 193280     (479 samples)
        * 99.0th: 360960     (107 samples)
          99.9th: 492032     (11 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 20 (s) (21 total samples)
          20.0th: 57         (5 samples)
        * 50.0th: 59         (7 samples)
          90.0th: 61         (9 samples)
          min=55, max=61
current rps: 60.97
Wakeup Latencies percentiles (usec) runtime 30 (s) (1815 total samples)
          50.0th: 69         (530 samples)
          90.0th: 2010       (725 samples)
        * 99.0th: 85376      (163 samples)
          99.9th: 375296     (17 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 30 (s) (1811 total samples)
          50.0th: 82048      (537 samples)
          90.0th: 193792     (721 samples)
        * 99.0th: 341504     (163 samples)
          99.9th: 492032     (16 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 30 (s) (31 total samples)
          20.0th: 58         (7 samples)
        * 50.0th: 60         (21 samples)
          90.0th: 60         (0 samples)
          min=55, max=61
current rps: 56.92
Wakeup Latencies percentiles (usec) runtime 40 (s) (2430 total samples)
          50.0th: 69         (712 samples)
          90.0th: 1914       (966 samples)
        * 99.0th: 85120      (219 samples)
          99.9th: 375296     (22 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 40 (s) (2424 total samples)
          50.0th: 81536      (712 samples)
          90.0th: 194816     (974 samples)
        * 99.0th: 341504     (214 samples)
          99.9th: 482816     (22 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 40 (s) (41 total samples)
          20.0th: 59         (22 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (15 samples)
          min=55, max=61
current rps: 58.94
Wakeup Latencies percentiles (usec) runtime 50 (s) (3042 total samples)
          50.0th: 68         (879 samples)
          90.0th: 1894       (1213 samples)
        * 99.0th: 85376      (274 samples)
          99.9th: 286208     (27 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 50 (s) (3035 total samples)
          50.0th: 81536      (906 samples)
          90.0th: 194304     (1207 samples)
        * 99.0th: 346624     (273 samples)
          99.9th: 480768     (27 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 50 (s) (51 total samples)
          20.0th: 59         (28 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 61         (23 samples)
          min=55, max=61
current rps: 60.98
Wakeup Latencies percentiles (usec) runtime 60 (s) (3650 total samples)
          50.0th: 68         (1070 samples)
          90.0th: 1874       (1450 samples)
        * 99.0th: 89216      (329 samples)
          99.9th: 286208     (33 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 60 (s) (3642 total samples)
          50.0th: 81536      (1101 samples)
          90.0th: 193792     (1442 samples)
        * 99.0th: 342528     (326 samples)
          99.9th: 480768     (33 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 60 (s) (61 total samples)
          20.0th: 59         (33 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (22 samples)
          min=55, max=61
current rps: 60.94
Wakeup Latencies percentiles (usec) runtime 70 (s) (4265 total samples)
          50.0th: 68         (1293 samples)
          90.0th: 1846       (1687 samples)
        * 99.0th: 89216      (384 samples)
          99.9th: 262656     (38 samples)
          min=1, max=383291
Request Latencies percentiles (usec) runtime 70 (s) (4259 total samples)
          50.0th: 81536      (1285 samples)
          90.0th: 193792     (1687 samples)
        * 99.0th: 348672     (382 samples)
          99.9th: 480768     (38 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 70 (s) (71 total samples)
          20.0th: 59         (37 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (27 samples)
          min=55, max=61
current rps: 60.97
Wakeup Latencies percentiles (usec) runtime 80 (s) (4872 total samples)
          50.0th: 68         (1437 samples)
          90.0th: 1922       (1927 samples)
        * 99.0th: 96896      (439 samples)
          99.9th: 349696     (44 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 80 (s) (4865 total samples)
          50.0th: 81536      (1468 samples)
          90.0th: 193792     (1938 samples)
        * 99.0th: 346624     (430 samples)
          99.9th: 484864     (44 samples)
          min=16122, max=536377
RPS percentiles (requests) runtime 80 (s) (81 total samples)
          20.0th: 59         (41 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (33 samples)
          min=55, max=61
current rps: 60.92
Wakeup Latencies percentiles (usec) runtime 90 (s) (5481 total samples)
          50.0th: 68         (1611 samples)
          90.0th: 1894       (2190 samples)
        * 99.0th: 89216      (494 samples)
          99.9th: 290304     (49 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 90 (s) (5474 total samples)
          50.0th: 81536      (1655 samples)
          90.0th: 193280     (2169 samples)
        * 99.0th: 348672     (489 samples)
          99.9th: 489984     (49 samples)
          min=16122, max=606819
RPS percentiles (requests) runtime 90 (s) (91 total samples)
          20.0th: 59         (47 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (36 samples)
          min=55, max=61
current rps: 59.94
Wakeup Latencies percentiles (usec) runtime 90 (s) (5482 total samples)
          50.0th: 68         (1611 samples)
          90.0th: 1906       (2191 samples)
        * 99.0th: 90496      (494 samples)
          99.9th: 290304     (49 samples)
          min=1, max=502160
Request Latencies percentiles (usec) runtime 90 (s) (5482 total samples)
          50.0th: 81536      (1658 samples)
          90.0th: 193280     (2173 samples)
        * 99.0th: 348672     (490 samples)
          99.9th: 489984     (49 samples)
          min=16122, max=606819
RPS percentiles (requests) runtime 90 (s) (91 total samples)
          20.0th: 59         (47 samples)
        * 50.0th: 59         (0 samples)
          90.0th: 60         (36 samples)
          min=55, max=61
average rps: 60.91
/benchmark/bin # 
```

</details>

Reply:
@jellllly420 Thanks for providing the sample output.

Now I understood why you would want to match the 28-th occurrences with the following configuration.

```json
    "result_extraction": {
        "search_pattern": "\\*",
        "nth_occurrence": 28,
    }
```

But having this magic number of 28 as part of the pattern matching rules is not good for _maintainability_. First, it is _unreadable_. Without seeing a sample output, people will have no idea why it is the 28-th occurence that is the result we want. Second, it is _fragile_. We could easily change our mind in the future to run for more time by adjusting the `-r` parameter. Every time we do so, we need to remember to adjust the `nth_occurrence`. This is exactly the reason why #1725 replaces all hardcoded IP addresses with regular expressions.

But if we don't introduce `nth_occurrence`, how we can match the result patterns that may appear in the output text for _multiple times_? Luckily, in this specific case of `schbench`, we don't have to.

```
-r (--runtime): How long to run before exiting (seconds, def: 30)
-i (--intervaltime): interval for printing latencies (seconds, def: 10) Print a report about our latencies and RPS ever N seconds.
```

We should change the command to specify the same time for both parameters. This means that the accumulated result will only show once.

```bash
RUNTIME=90
/benchmark/bin/schbench -r $RUNTIME -i RUNTIME -F 256 -n 5
```

Reply:
Thank you for the advice. Now I changed the run script like below
```sh
RUNTIME=90
REPORT_INTERVAL=$((RUNTIME+10))

/benchmark/bin/schbench -F 256 -n 5 -r $RUNTIME -i $REPORT_INTERVAL
```
Then schbench will only print a final summary after `$RUNTIME` seconds, which looks like below
```sh
Wakeup Latencies percentiles (usec) runtime 90 (s) (255474 total samples)
          50.0th: 7          (66640 samples)
          90.0th: 20         (68973 samples)
        * 99.0th: 272        (21770 samples)
          99.9th: 1266       (2301 samples)
          min=1, max=10481
Request Latencies percentiles (usec) runtime 90 (s) (255755 total samples)
          50.0th: 17120      (85972 samples)
          90.0th: 17568      (65273 samples)
        * 99.0th: 19616      (22817 samples)
          99.9th: 26464      (2249 samples)
          min=10794, max=48606
RPS percentiles (requests) runtime 90 (s) (9 total samples)
          20.0th: 2996       (3 samples)
        * 50.0th: 3004       (2 samples)
          90.0th: 3020       (4 samples)
          min=2982, max=3021
average rps: 3009.10
```
But still we cannot distinguish between the p99 wakeup latency and the p99 request latency with the current matching mechanism. We can (1) keep the `nth_occurrence` field (now it's stable, just 1/2/3, and won't be affected by any parameter change) or (2) reconstruct the result in run script like @grief8 did in this [comment](https://github.com/asterinas/asterinas/pull/1677#discussion_r1890120398)

The latest code adopted the first approach but I'm totally OK with switching to the second.

Reply:
I am fine with `nth_occurrence`, too.